platform/wayland/cog-platform-wl.c

```diff
@@ -132,16 +177,24 @@ struct shm_buffer {
 #endif
 
 #ifndef EGL_WL_create_wayland_buffer_from_image
-typedef struct wl_buffer * (EGLAPIENTRYP PFNEGLCREATEWAYLANDBUFFERFROMIMAGEWL) (EGLDisplay dpy, EGLImageKHR image);
+typedef struct wl_buffer *(EGLAPIENTRYP PFNEGLCREATEWAYLANDBUFFERFROMIMAGEWL)(EGLDisplay dpy, EGLImageKHR image);
 #endif
 
+typedef struct shell_operations {
```

Comment:
I think it might be good to be more consistent with naming - we have `shell_operations * shell_functions` just below here. Also, I notice that two extra xdg-related pointers are added to the `win_data` struct. Maybe it would be better to have a structure "shell_context" that encapsulates both the functions and any implementation-specific data necessary, and calling this "shell_protocol". Use of unions here might save some memory and make organisation a bit easier, for example:

```
typedef struct {
  void (*enter_fullscreen)();
  void (*exit_fullscreen)();
} ShellFunctions;

struct xdg_shell_protocol = {
  ShellFunctions functions;
  struct xdg_surface * xdg_surface;
  struct xdg_toplevel * xdg_toplevel;
};

struct f_shell_protocol {
  ShellFunctions functions;
  // ... Any data that fshell needs specifically ...
};

typedef union {
  ShellFunctions functions;
  struct xdg_shell_protocol xdg_context;
} ShellContext;

static ShellContext xdg_shell = {
  .functions = {
    &xdg_enter_fullscreen,
    &xdg_exit_fullscreen
  }
};
```

(excuse any incorrect syntax here, let's call this pseudo-code :))
So after this, you could have a `ShellContext* ctx` in win_data, and accessing functions would work much the same, like `win_data.ctx->functions.enter_fullscreen()` and individual data could be accessed like `win_data.ctx->xdg_context.xdg_surface`.